feat(prefer-single-binding): add rule

Olovyannikov opened the pull request:
Closes #116 

## Why is this important?

### Performance

Each `useUnit` call creates its own subscription management overhead. Combining them reduces:
- Number of hook calls
- Subscription management overhead
- Re-render coordination complexity

### Code clarity

A single `useUnit` call makes it easier to:
- See all dependencies at a glance
- Understand component's reactive logic
- Maintain and refactor code

### Example

```tsx
// correct
const Component = () => {
  const [value, setValue] = useUnit([$store, storeUpdated]);
}

// incorrect
const Component = () => {
  const [value] = useUnit([$store]);
  const [setValue] = useUnit([storeUpdated]);
}
```

## Separate stores and events calls

If you want to separate stores and events as independent groups, you can use option `separation`

When set to `allow`, allows separate `useUnit` calls for stores and events, but still enforces combining multiple calls within each group.

### Example

```tsx
// correct
const Component = () => {
  const [value, setValue] = useUnit([$store, storeUpdated]);
}

// if `separation` flag is 'enforce':
const Component = () => {
  const [value, setValue] = useUnit([$store, storeUpdated]);
  
   // will be transformed to:
   
  const [value] = useUnit([$store]);
  const [setValue] = useUnit([storeUpdated])
}

// also correct
const Component = () => {
  const [value] = useUnit([$store]);
  const [setValue] = useUnit([storeUpdated]);
}

// incorrect
const Component = () => {
  const [value] = useUnit([$store]);
  const [setValue] = useUnit([storeUpdated]);
  const [setAnotherValue] = useUnit([thirdEvent]);
}
```

Olovyannikov committed:
feat(prefer-single-binding): add rule

Olovyannikov committed:
feat(prefer-single-binding): add separate events and stores flag

Olovyannikov committed:
docs(prefer-single-binding): update docs

Olovyannikov committed:
feat(prefer-single-binding): update to v17

Olovyannikov committed:
fix(files): remove legacy empty files

Olovyannikov committed:
feat: add type-checker for detect unit type

Olovyannikov commented:
@kireevmp can you review this code please? 
I did it similarly to #175. I removed the unit type detection using heuristics and replaced it with type-check. And updated the test cases at the same time.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,407 @@
+    messages: {
+      multipleUseUnit:
+        "Multiple useUnit calls detected. Consider combining them into a single call for better performance.",
+      mixedStoresAndEvents:
+        "useUnit call contains both stores and events. Consider separating them into different calls.",
+    },
+    schema: [
+      {
+        type: "object",
+        properties: {
+          allowSeparateStoresAndEvents: { type: "boolean", default: false },
+          enforceStoresAndEventsSeparation: { type: "boolean", default: false },
+        },
+        additionalProperties: false,
+      },
+    ],
```
`allow` and `enforce` settings are not actually independent, because you can't enforce separation while also disallowing it. These are better merged into a string flag, i.e. `separation: "forbid" | "allow" | "enforce"`

Olovyannikov replied:
Yes, you're right. I made a string flag now.

kireevmp replied:
You added three items here: `forbid`, `allow` and `enforce` like I suggested. Can you explain how `allow` is different during code generation from `forbid` and `enforce` 'separation"? And how would the user know clearly why `allow` works this way?

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,407 @@
+        additionalProperties: false,
+      },
+    ],
+    fixable: "code",
```
The code fix output is not runtime equivalent to the input, so we shouldn't use autofix here. Since it modifies the runtime behavior, it must be a suggestion.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,407 @@
+      if (s.program) services = s
+    } catch {
+      services = null
+    }
```
The plugin generally requires TypeScript, and `eslint-typescript` guidelines recommend against rules that change behavior based on type information availability. So I suggest directly getting services as you need where you need without try/catch handling and other conditionals.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,407 @@
+
+      "FunctionDeclaration:exit": onFunctionExit,
+      "FunctionExpression:exit": onFunctionExit,
+      "ArrowFunctionExpression:exit": onFunctionExit,
```
Consider merging selectors into a single "logical" one `FunctionDeclaration, FunctionExpression, ArrowFunctionExpression`

Olovyannikov replied:
Indeed, you can simply list them separated by commas. Did it, thanks.

kireevmp commented on src/ruleset.ts:
```diff
@@ -13,6 +13,7 @@ const recommended = {
   "effector/no-unnecessary-duplication": "warn",
   "effector/no-useless-methods": "error",
   "effector/no-watch": "warn",
+  "effector/prefer-single-binding": "warn",
```
1. This is a `react` rule, not `recommended` one – it only concerns React users and not Vue users, for example
2. Since this is a stylistic rule (enforces code style, not correctness), we can't really include it into `react` preset either

Olovyannikov replied:
How about the `style` group?

kireevmp replied:
That'd be something close to `react-stylistic` preset, yeah. But I'm not sure we want to make a new preset just for one rule. Let's ponder on this once implementation is close to completion.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,407 @@
+  meta: {
+    type: "suggestion",
+    docs: {
+      description: "Recommend using a single useUnit call instead of multiple",
```
I'm a bit concerned with rule expectation mismatch here. It declares that it "recommends a single `useUnit` call", however

1. This rule does work with about non-destructuring calls `const a = useUnit($a)` and non-combining calls `const { a, b } = useUnit(model)` with property access
2. There are no tests illustrating how this rule works with `Effect`s
3. There are no tests that combines both array form and object form in a single component
4. There are no tests that play with variable positioning, for example where the unit consumed is not defined statically
   ```ts
   const { a } = useUnit({ a: $a })
   const { $b } = useContext(MyModel)
   const { b } = useUnit({ b: $b })
   ```
5. This rule does not declare how it works with `@@unitShape` (i.e. routers, queries)
6. With a proper setting, it actually enforces 2 `useUnit` calls with separation, so not a "single" call
7. I've also not found tests demonstrating how this rule works with aliasing 
   ```ts
   const { a } = useUnit({ a: $a });
   const { a: b } = useUnit({ a: $b })
   ```

My question here is – what are the expectations in these cases? Is there a sound "fix" we can provide for all of them? And should we actually check and report those cases?

Olovyannikov replied:
Yes, I will add more cases.

kireevmp replied:
Can we take a step back and work on this in comments first? It would be much easier to discuss conceptual solutions without diving into code other than examples. I'd love your opinions on what solutions, if any, are available for every listed case

Olovyannikov replied:
Aliasing works - I've added tests for this.
Also, we can't offer a fix for all cases; for such cases, we'll simply mark it as a warning.
Regarding effects, there's an issue suggesting creating a rule that prohibits working with effects directly. If that rule is included in the recommended rule, you can omit that case here, but by default, I currently separate calls when using separation enforce.

What about UnitShape - I'm not sure what the correct approach would be - highlight it as a warning? Or skip it, since it operates according to special rules?

Examples:

`separation: 'forbid'`

```tsx
// Store + Event
// ESLint: Multiple useUnit calls detected. Consider combining them into a single call for better performance. (effector/prefer-single-binding)

const products = useUnit(CartModel.$cart);
const onProductRemoveFromCart = useUnit(AddToCartModel.cartProductTotalRemoved);

```

```tsx
// Store + Event Array Shaped
// ESLint: Multiple useUnit calls detected. Consider combining them into a single call for better performance. (effector/prefer-single-binding)
// Suggest to merge this calls included

const [products] = useUnit([CartModel.$cartProducts]);
const [onProductRemoveFromCart] = useUnit([AddToCartModel.cartProductTotalRemoved]);
```

```tsx
// Store + Event Object Shaped
// ESLint: Multiple useUnit calls detected. Consider combining them into a single call for better performance. (effector/prefer-single-binding)
// Suggest to merge this calls included

const { products } = useUnit({
        products: CartModel.$cartProducts,
});
const { onProductRemoveFromCart } = useUnit({
      onProductRemoveFromCart: AddToCartModel.cartProductTotalRemoved,
});
```

```tsx
// Mixed shapes - Array + Object
// ESLint: Multiple useUnit calls detected. Consider combining them into a single call for better performance. (effector/prefer-single-binding)

const [products] = useUnit([CartModel.$cartProducts]);
const { onProductRemoveFromCart } = useUnit({
  onProductRemoveFromCart: AddToCartModel.cartProductTotalRemoved,
});
```

```tsx
// Using alias
// ESLint: Multiple useUnit calls detected. Consider combining them into a single call for better performance. (effector/prefer-single-binding)

const { products } = useUnitEffector({
        products: CartModel.$cartProducts,
    });
const [onProductRemoveFromCart] = useUnitEffector([AddToCartModel.cartProductTotalRemoved]);

```

```tsx
// named destructuring
// ESLint: Multiple useUnit calls detected. Consider combining them into a single call for better performance. (effector/prefer-single-binding)

const { a } = useUnitEffector({
        a: CartModel.$cartProducts,
    });
const { a: b } = useUnitEffector({
    a: AddToCartModel.cartProductTotalRemoved,
  });
```

kireevmp replied:
With both `@@unitShape` and a pre-build `useUnit(model)` where the argument is not a unit or shape of units, I'd suggest we fully exclude them from calculation. These can never be soundly "combined" so they're logically not subject to rule enforcement / separation.

---

With `Store + Event` basic case, this is actually the most common example of "invalid" code when developers duplicate `useUnit` calls without using destructuring, common copy-pasting. So we ought to provide a suggestion for this simplest case.

---

With `mixed shapes - Array + Object` case we can also provide a suggestion by treating 

```ts
const [products] = useUnit([CartModel.$cartProducts]);
// as equivalent to
const { products } = useUnit({ products: CartModel.$cartProducts });
```

and merging with other destructuring patterns.

This case will _always_ provide a suggestion unless either a rest property is used or variable names are duplicated in scope.

Using aliases adds a layer of complexity, however the accepting identifier is bound to be unique due to scoping rules, so

```ts
const { a: b } = useUnitEffector({ a: CartModel.$cartProducts });
// can be treated as
const { b } = useUnitEffector({ b: CartModel.$cartProducts });
```

for the sake of suggestion generation.

---

So the most notably here are the cases we aren't able to provide a suggestion:

1. `@@unitShape` and `useUnit(model)` – ignored by the rule due to unsound combination,
2. duplicate identifiers to prevent invalid JS from being generated, or
3. rest property and spread operator since we can't track `property -> unit` mapping.

Other cases should theoretically provide a valid suggestion.

kireevmp replied:
There's also a `useUnit` ordering problem since it can be applied to dynamically acquired `Unit`s, see example above with `useContext`. So a question is how would we treat the following code?

```ts
const a = useUnit($a) // (1)
const $b = useContext(MyModel) // (2)
const b = useUnit($b) // (3)
```

Should those be kept separate, suppressing the warning, or do we generate a suggestion for

```ts
const $b = useContext(MyModel)
const { a, b } = useUnit({ a: $a, b: $b })
```

And if we do so, how do we track if `a` has been used between `(1)` and `(3)`?

It seems even for basic cases this rule would also unfortunately require variable scope analysis to detect if units are declared in a current scope or a surrounding scope to prevent issues with temporal dead zone.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,407 @@
+  },
+} as const
+
+function getTypeFromChecker(node: Node.Expression, services: ParserServices): UnitType {
```
Let's use a shared typecheck helper `is` in this rule instead of custom checks?

Olovyannikov replied:
Cool! Found it, using it! :)

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,407 @@
+  }
+}
+
+function scoreTypes(types: UnitType[]): UnitType {
```
I'm not sure why we need to `score` something here and pre-group? Would it be possible to just collect the list of mappings `unit -> local ident` that is grouped later, with no regard of who was first or second, completely regenerating all `useUnit` calls?

Olovyannikov replied:
I agree, it looks unreliable. I'll see what I can do.

kireevmp reviewed (commented):
Hey @Olovyannikov! I skimmed through the PR and before we dive into actual implementation specifics, let's work out what this rule should check for and what we'd consider a violation. Right now it seems there are a few edge cases that need considering.

Olovyannikov committed:
refactor: add string flag, merge selectors, create new group 'style'

Olovyannikov committed:
feat(prefer-single-binding): detect local names doubles, tdz, aliases and plain calls

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+type Context = RuleContext<MessageIds, Options>
+type Fixer = RuleFixer
+type SourceCode = Context["sourceCode"]
+type Scope = ReturnType<SourceCode["getScope"]>
```
Why do we need to compute these aliases when plain imports are available? 

```
Fixer => RuleFixer
SourceCode => TSESLint.SourceCode
Scope => TSESLint.Scope.Scope
```

Let's strive to use built-ins as much as possible, without reinventing the wheel 🙏

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+    shape: "ObjectExpression.arguments",
+    list: "ArrayExpression.arguments",
+  },
+} as const
```
Why did you think to add `as const`? It does not seem to be relevant at all for type system here.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+/** Unit expression nodes of a call, regardless of whether it can be normalized (used for hoisting analysis). */
+function unitExpressionNodes(call: UseUnitCall): Node.Expression[] {
+  const argument = call.init.arguments[0]
+  if (!argument || argument.type === NodeType.SpreadElement) return []
```
Consider narrowing the selector on ESLint side to only include `useUnit` calls with `Object`, `Array` expressions or `Identifier` as first argument. May save couple of `if`s across the rule.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+    return argument.properties
+      .filter((p): p is Node.Property => p.type === NodeType.Property)
+      .map((p) => p.value as Node.Expression)
+  }
```
Let's avoid explicit type guards since TS can faithfully infer those now, modern style. These now only add extra noise.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+  if (argument.type === NodeType.ObjectExpression) {
+    return argument.properties
+      .filter((p): p is Node.Property => p.type === NodeType.Property)
+      .map((p) => p.value as Node.Expression)
```
Can we avoid a type cast? Looking at code, there's no guarantee made that non-expression arguments are excluded, so this is not sound.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+ */
+function extractBindings(call: UseUnitCall, sourceCode: SourceCode): NormalizedBinding[] | null {
+  const argument = call.init.arguments[0]
+  if (!argument || argument.type === NodeType.SpreadElement) return null
```
Same here, those ought to be pre-checked on ESQuery level if possible, or on top-level in visitor at the least. Helper routines are better when they work with clean, prepared data.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+  unit: string
+  /** unit expression node (for scope / hoisting analysis) */
+  unitNode: Node.Expression
+}
```
If I understand the context for the core functionality of this rule, we're only ever working with `(Identifier, Expression)` tuple. We're matching a "binding" in `Component` scope to an `Expression` that's passed to `useUnit`.

Therefore, the normalized form must only contain `{ unit: Expression, boundTo: Identifier }` shape – not strings and definitely not pre-computed source code. Minimal shape implies simpler extraction.

p.s. The source code of a node is only needed during reporting, so let's defer that work until we're sure there is a violation. Most code in checked is usually valid, so by deferring we improve "green run" performance.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+  if (targetForm === "array") {
+    return `const [${bindings.map((b) => b.local).join(", ")}] = ${calleeName}([${bindings.map((b) => b.unit).join(", ")}])`
+  }
+  return `const { ${bindings.map((b) => b.local).join(", ")} } = ${calleeName}({ ${bindings.map((b) => `${b.local}: ${b.unit}`).join(", ")} })`
```
A critical soundness issue here. This replaces the `.parent` of `VariableDeclarator` which is `VariableDeclaration`, and it can legally contain several declarations inside:

```ts
const a = useUnit($a), b = useUnit($b) // legal code
```

With this `.parent` access the first is "replaced" while the rest are literally erased.

What's worse is that the rule attempts to _remove_ the latter entry while simultaneously _replacing_ it with a fix, causing a crash due to overlap.

Another issue is that the rule does not respect `let` declarations, nor does it handle/ingest them properly.

A legal

```ts
let a = useUnit($a)

// ...

a ??= 0 // e.g. assigning a default
```

will jeopardize `useUnit` merging.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+      const variable = reference.resolved
+      if (!variable) continue
+      for (const def of variable.defs) {
+        if (def.node.range[0] >= anchorStart) return false
```
If I'm reading this correctly, this will basically disable the rule if a unit is just declared "after" the component. Consider

```tsx
function MyComponent() {
  const a = useUnit($a);
  const b = useUnit($b);

  return <span>{a + b}</span>
}

const $a = createStore(0)
const $b = createStore(1)
```

Legal code, `a` + `b` can be merged. But those are defined below `anchorStart` and erroneously fail this check.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+    if (unitExpressionNodes(call).every((node) => isHoistable(node, anchorStart, sourceCode))) subset.push(call)
+  }
+  return subset
+}
```
Q: Instead of retroactively checking every expression on if it can be hoisted, would it be better to just track where `Identifier` was defined while traversing the rule? I'm not sure what is easier: tracking all identifiers within a lexical scope or checking each one after to see hoisting?

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+  const removeEnd = nextChar === "\n" || nextChar === "\r" ? endIndex + 1 : endIndex
+
+  return fixer.removeRange([startIndex, removeEnd])
+}
```
How does this handle type annotations? Consider
```ts
const $a = createStore<1 | 2>(1) // narrow
// ...
const a: number = useUnit($a) // widens 1 | 2 => number
const b = useUnit($a) as number
```

Same with type casts, consider
```ts
const $a = useContext(MaybeStore) // => Store | undefined

const a = useUnit($a!) // <=
const b = useUnit($a as Store<number>) // <=
```

Are those erased and never regenerated? Or when are these preserved?

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+  return fixes
+}
+
+function generateSeparationFix(fixer: Fixer, call: UseUnitCall, context: Context, getNodeType: GetNodeType): RuleFix[] {
```
Why do we have three distinct traversals? `extractBindings`, `unitExpressionNodes` and `generateSeparationFix` – all of those ultimately re-implement the same traversal logic.

We have a primitive – a normalized binding. Let's use it extensively, instead of doing the work multiple times. Current situation causes unsoundness the code is **truly** hard to read and understand. This makes this rule by far the longest and hardest in this repo – but it shouldn't be, really.

There are a lot of edge cases in this rule. Given many code paths exist now, it is difficult to see all of them are handled. A brief look uncovered at least
- `enforce` mode not merging same-type `useUnit`
- dropping unknown key-value pairs
- statement write+remove overlap
- `const` + `let` conflict with latter not handled 
- `anchorStart` / `hoisting` is only computed wrt first `useUnit`, while valid fixes are available below
- `hoist` bug with after-component `Identifier` declarations
- valid `const { value: x, value: y } = useUnit({ value: $ })` causes `value: x` to be dropped

These, I believe, are not due to bad code in and of itself. I think these are all due to having done triple work in implementation. Those code paths need to be maximally consolidated if we are to resolve these.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+          `const { ${group.map(toKeyName).join(", ")} } = ${calleeName}({ ${group.map((p) => sourceCode.getText(p)).join(", ")} })`,
+      )
+
+    return [fixer.replaceText(call.statement, lines.join(`\n${indent}`))]
```
This seems to eventually just drop unknown properties from `useUnit`, potentially generating invalid _destructive_ code. Consider, for some reason, we couldn't detect type of a unit (let's say it resolved to `any`). Under no circumstances can we delete `useUnit($somethingThatIsAny)` – best-case we ought to filter it out and keep as-is, worst-case – do nothing with this `useUnit`.

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+              },
+            ],
+          }
+        : {}),
```
```suggestion
      suggest: mergeable
        ? [{ messageId: "multipleUseUnit", fix: (fixer) => generateMergeFix(fixer, captured, context) }]
        : undefined,
```

kireevmp commented on src/rules/prefer-single-binding/prefer-single-binding.ts:
```diff
@@ -0,0 +1,505 @@
+        if (!arg || arg.type === NodeType.SpreadElement) return
+        // Only non-destructuring calls over a single unit can be merged; a collection argument
+        // (`useUnit([...])` / `useUnit({...})`) or a non-unit argument (e.g. @@unitShape) is ignored.
+        if (arg.type === NodeType.ArrayExpression || arg.type === NodeType.ObjectExpression) return
```
Consolidate paths - if you're triggering on all calls after all, there's no reason to have three selectors if one suffices.

kireevmp commented on src/index.ts:
```diff
@@ -49,6 +50,7 @@ const base = {
     "no-useless-methods": noUselessMethods,
     "no-watch": noWatch,
     "prefer-useUnit": preferUseUnit,
+    "prefer-single-binding": preferSingleBinding,
```
Ordering
```suggestion
    "prefer-single-binding": preferSingleBinding,
    "prefer-useUnit": preferUseUnit,
```

kireevmp commented on src/ruleset.ts:
```diff
@@ -36,4 +36,8 @@ const future = {
-export const ruleset = { recommended, patronum, scope, react, future }
+const style = {
+  "effector/prefer-single-binding": "warn",
+} satisfies TSESLint.Linter.RulesRecord
```
Let's postpone adding a whole new `style` preset for now, please. Adding a preset is easy, removing not so much. I'd rather be cautious here, if you allow that.

kireevmp reviewed (commented):
Hey @Olovyannikov – thanks for taking the time to come back to polish. I see good progress here, I think there's a solid foundation now with normalization concept in place. What's left is to apply it consistently everywhere across the rule, to actually use that structure. The mental model here seems to be correct, which is great news.

If you're up to do another pass, I'd appreciate you also reviewing the test suite. Due to large rule size, under-employed normalization and ultimately duplicated code paths, the coverage is below 80% and the suite may not cover every corner case, something we really need here to be sure rule is sound.

Olovyannikov committed:
refactor(prefer-single-binding): rebuild around a single normalized binding

Collapse the three parallel traversals (extractBindings, unitExpressionNodes,
generateSeparationFix) into one analyze() pass that yields a single primitive:
an identifier bound to the unit expression passed to useUnit. Every helper now
consumes that structure, and source text is read only while building a fix.

Resolve the soundness issues raised in review:
- module units declared below the component are no longer a false TDZ hazard;
  only locals declared at/after the anchor block hoisting
- multi-declarator statements, let/var and type-annotated bindings are reported
  without a suggestion instead of producing a crashing or invalid fix
- the split fix keeps units whose type is undetermined instead of dropping them
- duplicate destructuring keys no longer silently drop a binding
- merging finds mergeable calls below one that cannot be hoisted
- enforce now merges same-type calls as well as splitting mixed ones

Drop the computed type aliases, the unsound `as Expression` cast and the
redundant `as const`/type guards in favour of the built-in ts-eslint types.

Tests cover the new edge cases; branch coverage rises from 79% to 89%.

Olovyannikov committed:
refactor(rules): drop the style preset and order the rule registry

prefer-single-binding stays available as effector/prefer-single-binding but is
no longer bundled in a preset — removing a preset after release is painful, so
hold off on adding one. Also order prefer-single-binding before prefer-useUnit.

Olovyannikov committed:
docs(prefer-single-binding): document separation modes and unmergeable cases

Add a comparison of forbid/allow/enforce code generation, spell out which calls
are ignored (@@unitShape, non-unit args, temporal dead zone) and which are
reported without a suggestion (multi-declarator, let, type annotations, etc.).
